### PR TITLE
generate tableOfContents field from HTML

### LINKED ISSUE #140: Include JSX headings in table of contents field
Right now, you can query an MDX node to get a table of contents that is built using the markdown headers in the MDX file. 

It would be really nice if this would also include headers rendered inside of JSX. I have scenarios where I want to put a header inside of a React component, but right now that will break my table of contents because `gatsby-mdx` only picks up Markdown headers.

### PR DESCRIPTION
**Pull request by chasemccoy**
This seems to work really well. I see two issues keeping this from shipping:

1. We are regenerating the HTML here instead of using a cached value, meaning that HTML for every page has to be generated twice (once for the `html` field, once for the `tableOfContents` field)
2. I am seeing the following error when running `yarn develop` on the kitchen sink example. Everything seems to work fine on subsequent queries though. 

```
  { Error: Cannot read property 'push' of undefined
      at defaultMergedResolver (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql-tools/dist/stitching/defaultMergedResolver.js:15:36)
      at field.resolve (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql-tools/dist/stitching/mergeSchemas.js:214:24)
      at resolveFieldValueOrError (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:531:18)
      at resolveField (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:495:16)
      at /Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:364:18
      at Array.reduce (<anonymous>)
      at executeFields (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:361:42)
      at collectAndExecuteSubfields (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:772:10)
      at completeObjectValue (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:762:10)
      at completeValue (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:660:12)
      at completeValueWithLocatedError (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:580:21)
      at completeValueCatchingError (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:556:21)
      at resolveField (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:497:10)
      at /Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:364:18
      at Array.reduce (<anonymous>)
      at executeFields (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:361:42)
      at collectAndExecuteSubfields (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:772:10)
      at completeObjectValue (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:762:10)
      at completeValue (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:660:12)
      at completeValueWithLocatedError (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:580:21)
      at completeValueCatchingError (/Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:556:21)
      at /Users/chasemccoy/Repositories/gatsby-mdx/node_modules/graphql/execution/execute.js:684:25
    message: 'Cannot read property \'push\' of undefined',
    locations: [ [Object] ],
    path: [ 'allMdx', 'edges', 5, 'node', 'tableOfContents' ] } ]
```

Once these issues are resolved this should fix #140

### CODE REVIEW

**Comment by chasemccoy** [comment]
The error seems to have something to do with the following line:

```js
const html = await withMDX(store.getState().webpack);
```

File: packages/gatsby-mdx/gatsby/extend-node-type.js [hunk 1]
```diff
@@ -198,7 +201,15 @@ module.exports = (
           }
         },
         async resolve(mdxNode, { maxDepth }) {
-          const { mdast } = await processMDX({ node: mdxNode });
+          const { body } = await processMDX({ node: mdxNode });
+          const withMDX = renderHTML(body);
+          const html = await withMDX(store.getState().webpack);
```
Comment (chasemccoy) [inline packages/gatsby-mdx/gatsby/extend-node-type.js]: The error seems to have something to do with this line.

**Comment by ChristopherBiscardi** [comment]
For 1) I don't really care. We can optimize later, lets get it working first and ship the working feature.

2) is something we should fix before shipping though

I'll take a look when I can, it looks like the docs fail too.

**Comment by ChristopherBiscardi** [comment]
The biggest problem is that we don't have the webpack config at the time that the toc is generated, which causes the error you're seeing. (the queries for pages happen before the site build happens). That's why it's failing and I don't know a way to fix that short of pre-running all of the webpack config collection somehow.

**Comment by ChristopherBiscardi** [comment]
going to close this since it's been hanging around awhile. The state is that if we're going to do this, we need to retrieve the webpack config much earlier than gatsby currently does (before the queries run, which is before the application builds)

**Pull request closed without merging**